diff -uNr a/makefile b/makefile
--- a/makefile	2019-03-21 10:28:57.989528000 +0800
+++ b/makefile	2019-03-21 10:31:13.775108600 +0800
@@ -74,6 +74,9 @@
 cross2:git_version
 	${cc_cross}   -o ${NAME}_cross    -I. ${SOURCES} ${FLAGS} -lrt -static -lgcc_eh -O3   
 
+cross2x:git_version
+	${cc_cross}   -o ${NAME}_cross    -I. ${SOURCES_AES_ACC} ${FLAGS} -lrt -static -lgcc_eh -O3 lib/aes_acc/asm/x64.S
+
 cross3:git_version
 	${cc_cross}   -o ${NAME}_cross    -I. ${SOURCES} ${FLAGS} -lrt -static -O3
 
